### ISSUE TITLE
Pasting text string with 40+ characters after asterisk freezes VS Code

### ISSUE THREAD

**Issue by jc-sanchez** [yzhang-gh/vscode-markdown#1117]
Issue Type: <b>Bug</b>

1. Open new VS Code - Insiders window with Markdown All in One enables.
2. Open new text file (example: press ctrl + n)
3. Change "Select language mode" to Markdown
4. Enter an asterisk (*)
5. Paste something that is 40+ characters long (ex: tttttttttttttttttttttttttttttttttttttttt)
    * **Expected:** Text in clipboard pastes and VS code keeps running
    * **Actual:** Text in clipboard pastes and VS code freezes

An alternative way to reproduce the bug it to input an astrick and then hold down a different key. After a while the editor will slow down and then freeze.

Extension version: 3.4.2
VS Code version: Code - Insiders 1.67.0-insider (e7c6ad74aff28965d0189ca1dd584f5e9829cf06, 2022-04-19T09:43:36.551Z)
OS version: Windows_NT x64 10.0.22598
Restricted Mode: No

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-1065G7 CPU @ 1.30GHz (8 x 1498)|
|GPU Status|2d_canvas: enabled<br>canvas_oop_rasterization: disabled_off<br>direct_rendering_display_compositor: disabled_off_ok<br>gpu_compositing: enabled<br>multiple_raster_threads: enabled_on<br>oop_rasterization: enabled<br>opengl: enabled_on<br>rasterization: enabled<br>raw_draw: disabled_off_ok<br>skia_renderer: enabled_on<br>video_decode: enabled<br>video_encode: enabled<br>vulkan: disabled_off<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|undefined|
|Memory (System)|15.60GB (4.16GB free)|
|Process Argv|--crash-reporter-id e352310c-b788-4383-91fa-2a2ced5d709a|
|Screen Reader|no|
|VM|0%|
</details><details>
<summary>A/B Experiments</summary>

```
vsliv695:30137379
vsins829:30139715
vsliv368cf:30146710
vsreu685:30147344
python383:30185418
vspor879:30202332
vspor708:30202333
vspor363:30204092
pythonvspyl392:30422396
pythontb:30258533
pythonptprofiler:30281269
vshan820:30294714
pythondataviewer:30285072
vscod805cf:30301675
pythonvspyt200:30323110
bridge0708:30335490
bridge0723:30353136
vsaa593cf:30376535
pythonvs932:30404738
wslgetstarted:30449409
pythonvsnew555:30438690
vscscmwlcmt:30465136
cppdebug:30475216
pynewfile477:30451556

```

</details>

**Comment by yzhang-gh** [comment]
Thanks for the detailed report.

I can confirm it is a bug after v3.4.0 and suspect it is related to #1092. Need to investigate.

**Comment by Lemmingh** [comment]
<https://github.com/yzhang-gh/vscode-markdown/blob/fe80c74e8aa2ad5e6f1b03aaac29f209cb4fbebb/src/extension.ts#L52>

Simplified:

```
(\p{Alphabetic}?\p{Alphabetic})*
```

Yes, ReDoS.

**Comment by yzhang-gh** [comment]
Interesting. Our regex simply add prefix `([*_]{1,2})?` and suffix `\1` to the VS Code's pattern

```
(\p{Alphabetic}|\p{Number}|\p{Nonspacing_Mark})(((\p{Alphabetic}|\p{Number}|\p{Nonspacing_Mark})|[_])?(\p{Alphabetic}|\p{Number}|\p{Nonspacing_Mark}))*
```

Don't know why VS Code comes up with such a pattern

![image](https://user-images.githubusercontent.com/7588612/164967740-9b780c34-72d2-4718-9738-673f41eaee24.png)

I can think of an alternative

```
(\p{Alphabetic}|\p{Number}|\p{Nonspacing_Mark})+(_(\p{Alphabetic}|\p{Number}|\p{Nonspacing_Mark})+)*
```

![image](https://user-images.githubusercontent.com/7588612/164967848-eace47b0-2bb5-4bdb-81d6-d2681fde2a4b.png)

and it seems to work well so far.

---

Also find that there are more corner cases in https://spec.commonmark.org/0.30/#emphasis-and-strong-emphasis although I don't see the necessity of dealing with them for now.

**Comment by Lemmingh** [comment]
This is a proven pattern, containing non-mutually exclusive tokens and usually denoted as `(AA|A)*`, which may open the door to catastrophic backtracking.

Perhaps the RegExp engine can run on a happy path there. While the backreference (`\1`) here adds a token after the quantified group, forcing a slow path to check all possible matches.

Anyway, reported at <https://github.com/microsoft/vscode/issues/148130>

**Comment by Lemmingh** [comment]
> ```javascript
> /([*_]{1,2})?(\p{Alphabetic}|\p{Number}|\p{Nonspacing_Mark})+(_(\p{Alphabetic}|\p{Number}|\p{Nonspacing_Mark})+)*\1/gu
> ```

Can be further simplified, IMO. It's not necessary to have so many branches and nested quantifiers, as it gives me a bad feeling of another proven pattern.

---

> there are more corner cases
> although I don't see the necessity of dealing with them for now

Of course. <https://github.com/yzhang-gh/vscode-markdown/issues/1092#issuecomment-1067999798>

It's not possible to express CM rules in a NFA.

**Comment by Lemmingh** [comment]
> Can be further simplified

At least, the alternation can be flattened to a character class, which eliminates extra steps especially during backtracking.

```javascript
/([*_]{1,2})?[\p{Alphabetic}\p{Number}\p{Nonspacing_Mark}]+(?:_[\p{Alphabetic}\p{Number}\p{Nonspacing_Mark}]+)*\1/gu
```

Reinforced, almost no backtracking:

```javascript
/([*_]{0,2})(?=(?<w>[\p{Alphabetic}\p{Number}\p{Nonspacing_Mark}]+(?:_[\p{Alphabetic}\p{Number}\p{Nonspacing_Mark}]+)*))\k<w>\1/gu
```

---

Performance test cases:

```javascript
"__" + "Ⅵ".repeat(100)
```

```javascript
"_" + "_ⅥⅥ".repeat(50)
```

**Comment by yzhang-gh** [comment]
Many thanks 👍. Will adopt this more efficient one when updating the `master` branch with the recent patch commits in `v3.4.0`.

**Comment by yzhang-gh** [comment]
has been fixed in v3.4.3
